### PR TITLE
Branch for Suzanne Review 02 July 

### PR DESCRIPTION
**Pull request by MatthewBurstein**

### CODE REVIEW

File: src/main/java/eventprocessing/GlobalConstants.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package eventprocessing;
+
+public class GlobalConstants {
+    public static int MULTIPLES_OF_MESSAGES = 10; //default: 10
+    public static long THIS_BUCKET_RANGE = (60 * 1000) - 1; //default: (60 * 1000) - 1
+    public static long BUCKET_UPPER_BOUND = 60 * 1000; //default: 60
+    public static int MAX_MESSAGE_DELAY_MINS = 5; //DO NOT CHANGE
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/GlobalConstants.java:7]: A scary comment like this should have an explanation so future devs can understand _why_ it shouldn't be changed. 🙂

File: src/main/java/eventprocessing/GlobalConstants.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package eventprocessing;
+
+public class GlobalConstants {
+    public static int MULTIPLES_OF_MESSAGES = 10; //default: 10
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/GlobalConstants.java:4]: What do the `default` comments mean in this class? What does it mean for a constant to have a default?

Comment (suzannehamilton) [inline src/main/java/eventprocessing/GlobalConstants.java:4]: Is this constant used anywhere?

File: src/main/java/eventprocessing/GlobalConstants.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+    public static long THIS_BUCKET_RANGE = (60 * 1000) - 1; //default: (60 * 1000) - 1
+    public static long BUCKET_UPPER_BOUND = 60 * 1000; //default: 60
+    public static int MAX_MESSAGE_DELAY_MINS = 5; //DO NOT CHANGE
+    public static long FIRST_LOOP_DURATION = MAX_MESSAGE_DELAY_MINS * BUCKET_UPPER_BOUND; //default: MAX_MESSAGE_DELAY_MINS * BUCKET_UPPER_BOUND
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/GlobalConstants.java:8]: These constants should all be marked as `final` so they can't be reassigned.

File: src/main/java/eventprocessing/Main.java [hunk 4]
```diff
@@ -1,69 +1,153 @@
...
+            Bucket removedBucket = bucketManager.removeExpiredBucket(System.currentTimeMillis());
+            if (removedBucket != null) {
+                csvFileService.writeBucketDataToFile(removedBucket);
+                bucketManager.createNextBucket();
+            }
+        }
+
+
+
+//        while (true) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/Main.java:111]: Commented-out code.

It's worth looking through a PR before submitting it for review to try to spot things like this before the reviewer does. 🙂

File: src/main/java/eventprocessing/amazonservices/S3Client.java [hunk 5]
```diff
@@ -8,32 +8,31 @@
...
 
         logger.debug("AWS S3 connection established");
 
         return s3;
     }
 
-    public S3ObjectInputStream generateS3InputStream() {
+    S3ObjectInputStream generateS3InputStream() {
         AmazonS3 s3 = clientBuilder();
-        S3ObjectInputStream s3is = null;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Client.java:28]: 👍

File: src/main/java/eventprocessing/amazonservices/S3Client.java [hunk 6]
```diff
@@ -8,32 +8,31 @@
...
         try {
             S3Object o = s3.getObject(S3Details.s3BucketLocation, S3Details.s3Key);
-            s3is = o.getObjectContent();
+            return o.getObjectContent();
         } catch (AmazonServiceException e) {
-            System.err.println(e.getErrorMessage());
-            System.exit(1);
+            logger.fatal("Fatal error retrieving locations.json from S3 Bucket - " + e.getErrorMessage());
+            throw new AmazonServiceException(e.getMessage());
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Client.java:33]: What about getting rid of this `catch` block entirely? It throws the same type of exception that it catches, so it doesn't really add anything. Or is the problem that the original exception doesn't include any information about what you were trying to fetch from S3?

File: src/main/java/eventprocessing/amazonservices/S3Client.java [hunk 7]
```diff
@@ -8,32 +8,31 @@
...
 
         return s3;
     }
 
-    public S3ObjectInputStream generateS3InputStream() {
+    S3ObjectInputStream generateS3InputStream() {
         AmazonS3 s3 = clientBuilder();
-        S3ObjectInputStream s3is = null;
         try {
             S3Object o = s3.getObject(S3Details.s3BucketLocation, S3Details.s3Key);
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/amazonservices/S3Client.java:29]: Watch out for single-letter names: `o` is a bit undescriptive. This isn't just a generic S3 object - it contains the locations data, because that's configured in the `getObject` parameters.

File: src/main/java/eventprocessing/analysis/Analyser.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package eventprocessing.analysis;
+
+import eventprocessing.models.Bucket;
+
+public class Analyser {
+
+    private static double reduceValues(double currentValue, double accumulator) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/analysis/Analyser.java:7]: Is this used?

File: src/main/java/eventprocessing/analysis/Analyser.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        return accumulator + currentValue;
+    }
+
+    public double getAverageValue(Bucket bucket) {
+
+        double total = bucket.getMessageValues()
+                .stream()
+                .reduce(0.0, Double::sum);
+
+        return total / bucket.getSqsResponse().size();
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/analysis/Analyser.java:17]: There's a code smell here - `Analyser` has to call two methods on `Bucket` to perform its calculation. This is called "feature envy", and it suggests that the data (in this case the values and SQS message count) and the operation (calculating the average) are too far apart.

What about bringing them together by moving this method onto `Bucket`?

File: src/main/java/eventprocessing/fileservices/CSVFileService.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+
+public class CSVFileService {
+    private final String outputCsvFile;
+    private final Analyser analyser;
+    private int ESTIMATED_LINE_LENGTH = 70;
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/fileservices/CSVFileService.java:17]: Should also be `final`?

File: src/main/java/eventprocessing/fileservices/CSVFileService.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+    public CSVFileService(String outputCsvFile) {
+        this.outputCsvFile = outputCsvFile;
+        this.analyser = new Analyser();
+    }
+
+    public void writeBucketDataToFile(Bucket bucketToWriteToFile) throws IOException {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/fileservices/CSVFileService.java:24]: This method appears to be called from the test, but the method below is the one that's actually used in `Main`, so we're not testing the real functionality. Could they be merged into the same method?

File: src/main/java/eventprocessing/models/Bucket.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+    public Bucket() {
+        //Constructor for InitialBucket
+    }
+
+    public void addResponse(SqsResponse sqsResponse) {
+        this.sqsResponse.add(sqsResponse);
+    }
+
+    public List<SqsResponse> getSqsResponse() {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/models/Bucket.java:28]: Nitpick: this method and the field it returns should be `SqsResponses` because it's a collection rathert than a single response.

File: src/main/java/eventprocessing/models/BucketManager.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        return remove(removedBucket);
+    }
+
+
+    public Bucket remove(Bucket bucket) {
+        buckets.remove(bucket);
+        return bucket;
+    }
+
+    public List<String> getMessageIds() {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/models/BucketManager.java:82]: Is this method necessary? It's only called by a test.

File: src/main/java/eventprocessing/models/BucketManager.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
+package eventprocessing.models;
+
+import eventprocessing.GlobalConstants;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class BucketManager {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/models/BucketManager.java:10]: This is another class where it would be useful to think about how the data and operations could be combined.

This class is called `BucketManager` but it doesn't do a lot of managing - it gets all of its instructions from outside the class, in `Main`.

How could you refactor the app so that `BucketManager` took responsibility for deciding when to roll over to the next bucket? It would move a lot of complicated logic out of `Main` and simplify the API of this class.

File: src/main/java/eventprocessing/models/SensorList.java [hunk 15]
```diff
@@ -1,16 +1,16 @@
...
-
-    public List<Sensor> getSensors() {
-        return sensors;
-    }
-}
+package eventprocessing.models;
+
+import java.util.List;
+
+public class SensorList {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/models/SensorList.java:5]: This class doesn't do anything except wrap `List<Sensor>`. Sometimes wrapping a type is useful because you can give it a more domain-specific name, but `SensorList` is pretty much the same thing as `List<Sensor>` so perhaps it would be simpler to remove it?

File: src/main/java/eventprocessing/models/SqsResponse.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public String toString() {
+        if (message != null) {
+            return "messageId: " + messageId + "\nMessage: " + String.valueOf(message.getLocationId());
+        } else {
+            return "messageId: " + messageId + "\nMessage: NULL";
+        }
+    }
+
+    public void setMessage(Message message) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/models/SqsResponse.java:28]: I just saw your comment on https://github.com/MatthewBurstein/EventProcessing/pull/1#discussion_r198131503 - I still think that having a `setMessage` method is hard to follow. If you need to deserialize the AWS response and the message separately then that's OK, but trying to combine them with a `setter` is confusing.

File: src/main/java/eventprocessing/models/SensorList.java (hunk 15 above)
Comment (suzannehamilton) [reply src/main/java/eventprocessing/models/SensorList.java:5]: Having said that, there's a `isWorkingSensor` method below which _would_ make sense to move onto `SensorList`! So maybe you shouldn't delete it - maybe it's worth keeping around but pushing some functionality onto it?

File: src/main/java/eventprocessing/responseservices/ResponseProcessor.java [hunk 17]
```diff
@@ -1,24 +1,43 @@
...
+import eventprocessing.models.SqsResponse;
+import eventprocessing.models.Bucket;
+import eventprocessing.models.SensorList;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class ResponseProcessor {
+    private final static Logger logger = LogManager.getLogger("ResponseProcessor");
+
+    public boolean isWorkingSensor(SqsResponse sqsResponse, SensorList sensorList) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseProcessor.java:13]: See my comment above about moving this method onto `SensorList`.

File: src/main/java/eventprocessing/responseservices/ResponseProcessor.java [hunk 18]
```diff
@@ -1,24 +1,43 @@
...
+                .stream()
+                .filter(sensor -> sensor.getId().equals(sqsResponse.getMessage().getLocationId()))
+                .count();
+        return count > 0;
+    }
+
+    public boolean isValidMessage(SqsResponse sqsResponse, SensorList sensorList, BucketManager bucketManager) {
+        boolean isValid = true;
+        for(Bucket bucket : bucketManager.getBuckets()) {
+            if(!isWorkingSensor(sqsResponse, sensorList)) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseProcessor.java:24]: Does this need to be checked for every bucket?

File: src/main/java/eventprocessing/responseservices/ResponseProcessor.java [hunk 19]
```diff
@@ -1,24 +1,43 @@
...
+                isValid = false;
+            }
+            if (isDuplicateMessage(sqsResponse, bucket)) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseProcessor.java:27]: This looks like more functionality that coule be pushed onto `BucketManager`.

File: src/main/java/eventprocessing/responseservices/ResponseProcessor.java [hunk 20]
```diff
@@ -1,24 +1,43 @@
...
+                isValid = false;
+            }
+        }
+        return isValid;
+    }
+
+    public boolean isValidMessage(SqsResponse sqsResponse, SensorList sensorList, Bucket bucket) {
```
Comment (suzannehamilton) [inline src/main/java/eventprocessing/responseservices/ResponseProcessor.java:34]: It's a bit suspicious that you have two different methods `isValidMessage` which are used in subtley differerent ways - one taking a `Bucket` and one taking a `BucketManager`. Hopefully pushing more behaviour onto `BucketManager` will help you clean this up.

**Review by suzannehamilton — COMMENTED** [review]
The names and the error handling are looking much better. 👍 

My main comments this time are about encapsulation, and thinking about pushing more functionality into classes so that the data and operations aren't so separate.

I've just skimmed the tests, because I think they'll change a lot when you refactor the app code. Have a think about how big your tests are getting, though - if you're finding them frustrating to work with, it'd be good to have a chat about why and what we can do about it.

File: src/main/java/eventprocessing/amazonservices/S3Client.java (hunk 6 above)
Comment (MatthewBurstein) [reply src/main/java/eventprocessing/amazonservices/S3Client.java:33]: If we don't catch the error here then we can't log anything about what caused it and we have onyl the stack trace in the console to work with. Is this not the correct approach?

File: src/main/java/eventprocessing/fileservices/CSVFileService.java (hunk 11 above)
Comment (MatthewBurstein) [reply src/main/java/eventprocessing/fileservices/CSVFileService.java:24]: Not quite sure what you mean here - we use both, one when parsing the list of buckets at the start of the program, and one which runs regularly once the program has been running for 5 minutes.  Also both are called in the test (although it is probably not that clear what it would look like if either/both fail...

File: src/main/java/eventprocessing/amazonservices/S3Client.java (hunk 6 above)
Comment (suzannehamilton) [reply src/main/java/eventprocessing/amazonservices/S3Client.java:33]: It's a good rule of thumb that you should either log an error or throw it, but not both. If you do both, you end up recording the same error twice in two different places, which usually makes it _harder_ to debug the issue.

I think the best thing to do here depends on how much useful information there is in the original error message.

If it already contains the S3 key "locations.json", then your log doesn't add anything, so don't bother with the catch block - let the original error bubble up, and it'll be logged in the console when the app crashes.

If it doesn't, then throw another exception which includes your helpful message and the _whole_ of the original exception so the console error includes the full stack trace.
